### PR TITLE
Update plugin for Elasticsearch 5.3.0

### PR DESCRIPTION
**Pull request by lovemeblender**

### CODE REVIEW

File: src/main/java/org/elasticsearch/module/WordDelimiterRunnable.java [hunk 1]
```diff
@@ -43,7 +44,6 @@ protected void doRun() {
     WordDelimiterActionListener listener = WordDelimiterActionListener.getInstance();
     SearchRequest searchRequest = client.prepareSearch().setSearchType(SearchType.QUERY_THEN_FETCH)
         .setIndices(index).setTypes(type).setSize(RESULTS_SIZE).request();
-    searchRequest.listenerThreaded(false);
```
Comment (bill-kolokithas) [inline src/main/java/org/elasticsearch/module/WordDelimiterRunnable.java:46]: Commit body should reference the upstream commit that introduced this change.
https://github.com/elastic/elasticsearch/commit/b87d360e79726813db90c7acecc965057a4a39ed

File: src/main/java/org/elasticsearch/module/WordDelimiterRunnable.java [hunk 2]
```diff
@@ -25,9 +25,9 @@
 
   public WordDelimiterRunnable(Client client, Settings settings) {
     this.client = client;
-    this.index = settings.get("plugin.dynamic_word_delimiter.protected_words_index", INDEX_NAME);
-    this.type = settings.get("plugin.dynamic_word_delimiter.protected_words_type", INDEX_TYPE);
-    this.interval = settings.getAsTime("plugin.dynamic_word_delimiter.refresh_interval", REFRESH_INTERVAL).getMillis();
+    this.index = settings.get("node.attr.plugin.dynamic_word_delimiter.protected_words_index", INDEX_NAME);
```
Comment (bill-kolokithas) [inline src/main/java/org/elasticsearch/module/WordDelimiterRunnable.java]: Don't forget to change this in the Readme as well.

File: src/main/java/org/elasticsearch/module/WordDelimiterRunnable.java [hunk 3]
```diff
@@ -14,7 +14,7 @@
   public static final TimeValue REFRESH_INTERVAL = TimeValue.timeValueMinutes(5);
   public static final String INDEX_NAME = "protected_words";
   public static final String INDEX_TYPE = "word";
-  public static final int RESULTS_SIZE = 150000;
+  public static final int RESULTS_SIZE = 10000;
```
Comment (bill-kolokithas) [inline src/main/java/org/elasticsearch/module/WordDelimiterRunnable.java:17]: Upstream commit: https://github.com/elastic/elasticsearch/commit/e4981968adb045770b93ace12c175a6ba756853a

I also see a lot of tests specifying the size as zero. The default is 10, but a value of zero might default to the `index.max_result_window`, but i'm not sure about this one.

File: src/main/java/org/elasticsearch/plugin/WordDelimiterPlugin.java [hunk 4]
```diff
@@ -1,37 +1,40 @@
...
     services.add(WordDelimiterService.class);
   }
 
-  public String name() {
-    return "dynamic-word-delimiter";
+  @Override
+  public Collection<Class<? extends LifecycleComponent>> getGuiceServiceClasses() {
+    Collection<Class<? extends LifecycleComponent>> services = new ArrayList<>();
+    services.add(WordDelimiterService.class);
+    return services;
```
Comment (bill-kolokithas) [inline src/main/java/org/elasticsearch/plugin/WordDelimiterPlugin.java]: I would use a singleton to cut the verbosity down like so:
`return Collections.singleton(WordDelimiterService.class);`

File: src/main/java/org/elasticsearch/plugin/WordDelimiterPlugin.java [hunk 5]
```diff
@@ -1,37 +1,40 @@
...
+          new ArrayList();
 
   public WordDelimiterPlugin() {
     services.add(WordDelimiterService.class);
   }
 
-  public String name() {
-    return "dynamic-word-delimiter";
+  @Override
+  public Collection<Class<? extends LifecycleComponent>> getGuiceServiceClasses() {
```
Comment (bill-kolokithas) [inline src/main/java/org/elasticsearch/plugin/WordDelimiterPlugin.java]: Can you provide a source for the new way of registering services?

File: src/main/java/org/elasticsearch/module/WordDelimiterRunnable.java (hunk 3 above)
Comment (lovemeblender) [reply src/main/java/org/elasticsearch/module/WordDelimiterRunnable.java:17]: Added reference to upstream commit

File: src/main/java/org/elasticsearch/module/WordDelimiterRunnable.java (hunk 2 above)
Comment (lovemeblender) [reply src/main/java/org/elasticsearch/module/WordDelimiterRunnable.java]: Updated readme

File: src/main/java/org/elasticsearch/plugin/WordDelimiterPlugin.java (hunk 5 above)
Comment (lovemeblender) [reply src/main/java/org/elasticsearch/plugin/WordDelimiterPlugin.java]: I didn't find any documentation on this. We're supposed to deduce that from the API I guess, as it was the only method around services in Plugin class.

File: src/main/java/org/elasticsearch/module/WordDelimiterRunnable.java (hunk 1 above)
Comment (lovemeblender) [reply src/main/java/org/elasticsearch/module/WordDelimiterRunnable.java:46]: Added reference to upstream commit

File: src/main/java/org/elasticsearch/plugin/WordDelimiterPlugin.java [hunk 6]
```diff
@@ -1,37 +1,53 @@
...
-  @SuppressWarnings("rawtypes")
   @Override
-  public Collection<Class<? extends LifecycleComponent>> services() {
-    return services;
+  public List<Setting<?>> getSettings() {
+
+    List<Setting<?>> settings = Arrays.asList(
+            new Setting<>(
+                    "plugin.dynamic_word_delimiter.protected_words_index",
+                    "protected_words",
```
Comment (bill-kolokithas) [inline src/main/java/org/elasticsearch/plugin/WordDelimiterPlugin.java:43]: Since we provide a `fallbackSetting` here, we might not need to set the default values in WordDelimiterRunnable.java.
Remove the static variables and use the [get/1](https://github.com/elastic/elasticsearch/blob/v5.3.0/core/src/main/java/org/elasticsearch/common/settings/Settings.java#L245) version.

Commit that introduced the new way or registering settings: https://github.com/elastic/elasticsearch/commit/b22c526b34c5bba261ec79a083a41e40f6f1483d

Comment (lovemeblender) [reply src/main/java/org/elasticsearch/plugin/WordDelimiterPlugin.java:43]: Discussed about that. Plain get won't work for the refresh interval as it returns a string. Agreed to maintain a single point where we initialize these constants, which is going to be WordDelimiterRunnable class.

File: src/main/plugin-metadata/plugin-security.policy [hunk 7]
```diff
@@ -0,0 +1,7 @@
+grant {
```
Comment (lovemeblender) [inline src/main/plugin-metadata/plugin-security.policy:1]: Note to current and future reviewers: we need to take a decision on whether we are going to keep the security permissions or not. The reason they appear is that they were needed by the tests and therefore made the assumption that would be needed in a production environment as well. However removing them and trying the main manual testing flow as described at the readme of the repo works fine which brings us to the conclusion that probably these are needed by the tests only.

A possible solution would be to remove them and keep the security manager disabled when testing (as it is right now). Otherwise we could play safe and keep them, just in case something unpredictable comes up in production with the risk of:
1. Having a potential security vulnerability
2. Not prompting when the security permissions are asking for it and causing the installation to hang
